### DISCUSSION TITLE
Implementing a non-Diesel based database (FoundationDB) to vaultwarden. Got questions and need advice

### DISCUSSION THREAD

**Discussion by alvinpeters** [dani-garcia/vaultwarden#4705] [General]
## TL;DR. What are the questions then?

- Would this feature be worth upstreaming? I'd be down to maintain this part if this gets accepted.
- If the answer to above is yes/maybe, should I submit a draft PR now so everyone can easily check on it?
- What do I do with a single unsafe code block and the linter being set to forbid everything unsafe?
- How do I fit the current schema into a key-value store?
- If I decide to store large tables into a single value. Which serialisation protocol do I go with?
JSON, BSON, Protobuf, Cap'n Proto, or do I get to choose whatever I want as long as it's easy to maintain?
- Is my current approach of making new traits and pool for non-diesel database good enough?

## Background

Hey! Apologies in advance for my possibly incoherent writing as it's kinda my first time adding a major feature to other people's code.

I want to use vaultwarden in my HA availability cluster, but don't want to install another database backend since I already have FoundationDB which is quite heavy enough, so I set on trying to implement FoundationDB with [foundationdb-rs](https://github.com/foundationdb-rs/foundationdb-rs) crate.

You can check out my progress here: https://github.com/alvinpeters/vaultwarden/tree/add-foundationdb-setup-nondiesel
Should I submit a draft PR already?

This is turning to be way harder than I expected. as the existing schemas are heavily reliant on SQL table structure. I'm figuring out the best way to store them as key-value pairs. I reckon I can follow [this table design recipe](https://apple.github.io/foundationdb/tables.html), but it would end up storing too much copies of the same keys (I guess it's literally unavoidable in k-v stores but a key per column is still excessive.)

I'm considering storing large tables that are often fetched together in a single key-value pair and call it day. Only having separate lookup subspaces for values that need to be frequently looked up. I'm considering storing blobs using either JSON (so you can directly chuch it to the API after fetching), BSON, Protobuf, or Cap'n Proto.

Oh yeah and foundationdb crate requires the use of [a single unsafe code](https://docs.rs/foundationdb/latest/foundationdb/fn.boot.html) just to start up and then ensure that it gets dropped once on shutdown. Since the crate forbids unsafe code for now. How do I handle it? Do I set the linter to only warn so people don't get warned unless they enable the fdb feature?

And don't get me started on FoundationDB not being able to use diesel. I've refactored the macros already, and I'm writing custom Pool implementation, connection traits, and even schema macros so it's easy to slap it into the db_run and db_object macros once done. This should also allow other NoSQL databases like Redis/Valkey and Cassandra to be implemented down the line. Is there a better way to go with this or what I'm doing is good enough?

Cheers for answers!
